From 5e702cd23a359cc5dfa6ff0f45101f9b89041f2f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Danis?= <frederic.danis@collabora.com>
Date: Mon, 27 Nov 2023 13:07:31 +0100
Subject: [PATCH] Bluez5: backend-native: HSP AG release SCO link on
 AT+CKPD=200
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Bluetooth PTS test HSP/AG/ACR/BV-01-I request AG to release the SCO link
upon reception of AT+CKPD=200 reception
---
 spa/plugins/bluez5/backend-native.c |  5 +++--
 spa/plugins/bluez5/bluez5-device.c  | 22 ++++++++++++++++++++++
 spa/plugins/bluez5/defs.h           |  4 ++++
 3 files changed, 29 insertions(+), 2 deletions(-)

diff --git a/spa/plugins/bluez5/backend-native.c b/spa/plugins/bluez5/backend-native.c
index 8144b09fa..51331a68d 100644
--- a/spa/plugins/bluez5/backend-native.c
+++ b/spa/plugins/bluez5/backend-native.c
@@ -442,7 +442,7 @@ static void rfcomm_emit_volume_changed(struct rfcomm *rfcomm, int id, int hw_vol
 static bool rfcomm_hsp_ag(struct rfcomm *rfcomm, char* buf)
 {
 	struct impl *backend = rfcomm->backend;
-	unsigned int gain, dummy;
+	unsigned int gain;
 
 	/* There are only three HSP AT commands:
 	 * AT+VGS=value: value between 0 and 15, sent by the HS to AG to set the speaker gain.
@@ -465,8 +465,9 @@ static bool rfcomm_hsp_ag(struct rfcomm *rfcomm, char* buf)
 			rfcomm_send_reply(rfcomm, "ERROR");
 			spa_log_debug(backend->log, "RFCOMM receive unsupported VGM gain: %s", buf);
 		}
-	} else if (sscanf(buf, "AT+CKPD=%d", &dummy) == 1) {
+	} else if (spa_strstartswith(buf, "AT+CKPD=200") == 1) {
 		rfcomm_send_reply(rfcomm, "OK");
+		spa_bt_device_emit_switch_profile(rfcomm->device);
 	} else {
 		return false;
 	}
diff --git a/spa/plugins/bluez5/bluez5-device.c b/spa/plugins/bluez5/bluez5-device.c
index 8d6ef494b..595065785 100644
--- a/spa/plugins/bluez5/bluez5-device.c
+++ b/spa/plugins/bluez5/bluez5-device.c
@@ -1033,11 +1033,33 @@ static void device_connected(void *userdata, bool connected)
 		set_initial_profile(this);
 }
 
+static void device_switch_profile(void *userdata)
+{
+	struct impl *this = userdata;
+	uint32_t profile;
+
+	switch(this->profile) {
+	case DEVICE_PROFILE_OFF:
+		profile = DEVICE_PROFILE_HSP_HFP;
+		break;
+	case DEVICE_PROFILE_HSP_HFP:
+		profile = DEVICE_PROFILE_OFF;
+		break;
+	default:
+		return;
+	}
+
+	spa_log_debug(this->log, "%p: device switch profile %d -> %d", this, this->profile, profile);
+
+	set_profile(this, profile, 0, false);
+}
+
 static const struct spa_bt_device_events bt_dev_events = {
 	SPA_VERSION_BT_DEVICE_EVENTS,
 	.connected = device_connected,
 	.codec_switched = codec_switched,
 	.profiles_changed = profiles_changed,
+	.switch_profile = device_switch_profile,
 };
 
 static int impl_add_listener(void *object,
diff --git a/spa/plugins/bluez5/defs.h b/spa/plugins/bluez5/defs.h
index 81e51640f..fbe0eae1a 100644
--- a/spa/plugins/bluez5/defs.h
+++ b/spa/plugins/bluez5/defs.h
@@ -455,6 +455,9 @@ struct spa_bt_device_events {
 	/** Profile configuration changed */
 	void (*profiles_changed) (void *data, uint32_t prev_profiles, uint32_t prev_connected);
 
+	/** Switch profile between OFF and HSP_HFP */
+	void (*switch_profile) (void *data);
+
 	/** Device freed */
 	void (*destroy) (void *data);
 };
@@ -533,6 +536,7 @@ void spa_bt_device_update_last_bluez_action_time(struct spa_bt_device *device);
 #define spa_bt_device_emit_connected(d,...)	        spa_bt_device_emit(d, connected, 0, __VA_ARGS__)
 #define spa_bt_device_emit_codec_switched(d,...)	spa_bt_device_emit(d, codec_switched, 0, __VA_ARGS__)
 #define spa_bt_device_emit_profiles_changed(d,...)	spa_bt_device_emit(d, profiles_changed, 0, __VA_ARGS__)
+#define spa_bt_device_emit_switch_profile(d)		spa_bt_device_emit(d, switch_profile, 0)
 #define spa_bt_device_emit_destroy(d)			spa_bt_device_emit(d, destroy, 0)
 #define spa_bt_device_add_listener(d,listener,events,data)           \
 	spa_hook_list_append(&(d)->listener_list, listener, events, data)
-- 
2.34.1

